cmd/tempo-vulture/main.go

```diff
@@ -507,40 +558,70 @@ func hasMissingSpans(t *tempopb.Trace) bool {
 
 func constructTraceFromEpoch(epoch time.Time) *tempopb.Trace {
 	r := newRand(epoch)
-	traceIDHigh := r.Int63()
-	traceIDLow := r.Int63()
+
+	info := &traceInfo{
+		timestamp:           epoch,
+		r:                   r,
+		traceIDHigh:         r.Int63(),
+		traceIDLow:          r.Int63(),
+		longWritesRemaining: r.Int63n(longWriteBatchChance),
+	}
 
 	trace := &tempopb.Trace{}
 
-	for i := int64(0); i < generateRandomInt(1, 100, r); i++ {
-		batch := makeThriftBatch(traceIDHigh, traceIDLow, r, epoch)
-		internalTrace := jaegerTrans.ThriftBatchToInternalTraces(batch)
-		conv, err := internalTrace.ToOtlpProtoBytes()
-		if err != nil {
-			logger.Error(err.Error())
-		}
+	addBatches := func(t *traceInfo, trace *tempopb.Trace) {
+		for i := int64(0); i < generateRandomInt(1, batchHighWaterMark, r); i++ {
+			batch := makeThriftBatch(t.traceIDHigh, t.traceIDLow, r, epoch)
+			internalTrace := jaegerTrans.ThriftBatchToInternalTraces(batch)
+			conv, err := internalTrace.ToOtlpProtoBytes()
+			if err != nil {
+				logger.Error(err.Error())
+			}
 
-		t := tempopb.Trace{}
-		err = t.Unmarshal(conv)
-		if err != nil {
-			logger.Error(err.Error())
-		}
+			t := tempopb.Trace{}
+			err = t.Unmarshal(conv)
+			if err != nil {
+				logger.Error(err.Error())
+			}
 
-		// Due to the several transforms above, some manual mangling is required to
-		// get the parentSpanID to match.  In the case of an empty []byte in place
-		// for the ParentSpanId, we set to nil here to ensure that the final result
-		// matches the json.Unmarshal value when tempo is queried.
-		for _, b := range t.Batches {
-			for _, l := range b.InstrumentationLibrarySpans {
-				for _, s := range l.Spans {
-					if len(s.GetParentSpanId()) == 0 {
-						s.ParentSpanId = nil
+			// Due to the several transforms above, some manual mangling is required to
+			// get the parentSpanID to match.  In the case of an empty []byte in place
+			// for the ParentSpanId, we set to nil here to ensure that the final result
+			// matches the json.Unmarshal value when tempo is queried.
+			for _, b := range t.Batches {
+				for _, l := range b.InstrumentationLibrarySpans {
+					for _, s := range l.Spans {
+						if len(s.GetParentSpanId()) == 0 {
+							s.ParentSpanId = nil
+						}
 					}
 				}
 			}
+
+			trace.Batches = append(trace.Batches, t.Batches...)
+		}
+	}
+
+	addBatches(info, trace)
+
+	lastWrite := info.timestamp
+
+	for info.longWritesRemaining > 0 {
+		info.longWritesRemaining--
+
+		lastWrite = lastWrite.Add(tempoLongWriteBackoffDuration)
+
+		// If the last write has happened very recently, we'll wait a bit to make
+		// sure the write has taken place, and then add the batches that would have
+		// just been written.
+		if time.Since(lastWrite) < (1 * time.Second) {
```

Comment:
this is pretty janky.

from a seed we should be able to determine when a trace was completed and either query or not query the trace if we feel it is not complete.

putting sleeps here is going to weirdly impact the timings of other loops.

Reply:
Alright, I've washed off some of that jank.  No more sleeps to construct the expected trace, and we only search/query when we expect that all writes have taken place.